src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -0,0 +1,423 @@
+                var conversion = Conversions.ClassifyBuiltInConversion(inputType, type, ref discardedUseSiteDiagnostics);
+                if (input.Type.IsDynamic() ? type.SpecialType == SpecialType.System_Object : conversion.IsImplicit)
+                {
+                    // type test not needed, only the type cast
+                }
+                else
+                {
+                    // both type test and cast needed
+                    decisions.Add(new BoundTypeDecision(syntax, type, input));
+                }
+
+                var evaluation = new BoundDagTypeEvaluation(syntax, type, input);
+                input = new BoundDagTemp(syntax, type, evaluation, 0);
+                decisions.Add(evaluation);
+            }
+
+            return input;
+        }
+
+        private void MakeDecisionsAndBindings(
+            BoundDagTemp input,
+            BoundConstantPattern constant,
+            ArrayBuilder<BoundDagDecision> decisions,
+            ArrayBuilder<(BoundExpression, BoundDagTemp)> bindings,
+            ref HashSet<DiagnosticInfo> discardedUseSiteDiagnostics)
+        {
+            input = ConvertToType(input, constant.Syntax, constant.Value.Type, decisions, ref discardedUseSiteDiagnostics);
+            decisions.Add(new BoundValueDecision(constant.Syntax, constant.ConstantValue, input));
+        }
+
+        private void MakeDecisionsAndBindings(
+            BoundDagTemp input,
+            BoundRecursivePattern recursive,
+            ArrayBuilder<BoundDagDecision> decisions,
+            ArrayBuilder<(BoundExpression, BoundDagTemp)> bindings,
+            ref HashSet<DiagnosticInfo> discardedUseSiteDiagnostics)
+        {
+            Debug.Assert(input.Type == recursive.InputType);
+            NullCheck(input, recursive.Syntax, decisions);
+            if (recursive.DeclaredType != null && recursive.DeclaredType.Type != input.Type)
+            {
+                input = ConvertToType(input, recursive.Syntax, recursive.DeclaredType.Type, decisions, ref discardedUseSiteDiagnostics);
+            }
+
+            if (!recursive.Deconstruction.IsDefault)
+            {
+                // we have a "deconstruction" form, which is either an invocation of a Deconstruct method, or a disassembly of a tuple
+                if (recursive.DeconstructMethodOpt != null)
+                {
+                    var method = recursive.DeconstructMethodOpt;
+                    var evaluation = new BoundDagDeconstructEvaluation(recursive.Syntax, method, input);
+                    decisions.Add(evaluation);
+                    int extensionExtra = method.IsStatic ? 1 : 0;
+                    int count = Math.Min(method.ParameterCount - extensionExtra, recursive.Deconstruction.Length);
+                    for (int i = 0; i < count; i++)
+                    {
+                        var pattern = recursive.Deconstruction[i];
```

Comment:
>var [](start = 24, length = 3)

Please spell out the type.

Reply:
Resolved in [https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec](https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec) of [https://github.com/dotnet/roslyn/pull/23357](https://github.com/dotnet/roslyn/pull/23357)

---
In reply to: [156535684](https://github.com/dotnet/roslyn/pull/23209#discussion_r156535684) [](ancestors = 156535684)